diff --git a/userprog/exception.c b/userprog/exception.c
index fd70f99..74ec284 100644
--- a/userprog/exception.c
+++ b/userprog/exception.c
@@ -140,14 +140,19 @@ page_fault (struct intr_frame *f) {
 	write = (f->error_code & PF_W) != 0;
 	user = (f->error_code & PF_U) != 0;
 
-	if ((!not_present && write) || (fault_addr < 0x400000 || fault_addr >= USER_STACK))
+	if ((!not_present && write) || (fault_addr < 0x400000 || fault_addr >= USER_STACK)) {
 		exit_(-1);
+   }
 #ifdef VM
 	/* For project 3 and later. */
-	if (vm_try_handle_fault (f, fault_addr, user, write, not_present))
-		return;
+	if (vm_try_handle_fault (f, fault_addr, user, write, not_present)) {
+      return;
+   }
 #endif
 
+   if (user) {
+      exit_(-1);
+   }
 	/* Count page faults. */
 	page_fault_cnt++;

